### PR TITLE
Absolute swerve

### PR DESCRIPTION
**Pull request by OliverW10**
updated swerve to use absolute mag encoders on separate dedicated motor controllers

### CODE REVIEW

File: components/chassis.py [hunk 1]
```diff
@@ -57,6 +60,9 @@ def __init__(
             ctre.FeedbackDevice.IntegratedSensor, 0, 10
         )
 
+        self.encoder = encoder
+        self.encoder.setPosition(encoder_offset)
```
Comment (auscompgeek) [inline components/chassis.py:64]: Does this have an effect in absolute mode? I thought it only lets you set the number of wraparounds.

Comment (OliverW10) [reply components/chassis.py]: fixed, also added comments for what to change to for cancoders

File: components/chassis.py [hunk 2]
```diff
@@ -61,7 +60,9 @@ def __init__(
         )
 
         self.encoder = encoder
-        self.encoder.setPosition(encoder_offset)
+
+        # change to encoder.configMagnetOffset(encoder_offset)
+        self.encoder.talon.configIntegratedSensorOffset(encoder_offset)
```
Comment (auscompgeek) [inline components/chassis.py:65]: Does this method actually exist on a TalonSRX? Surely this refers to the Talon FX integrated sensor.

File: components/chassis.py [hunk 3]
```diff
@@ -61,7 +60,9 @@ def __init__(
         )
 
         self.encoder = encoder
-        self.encoder.setPosition(encoder_offset)
+
+        # change to encoder.configMagnetOffset(encoder_offset)
```
Comment (auscompgeek) [inline components/chassis.py:64]: I think we would want to do this outside of the code

File: components/chassis.py [hunk 4]
```diff
@@ -104,7 +111,9 @@ def set(self, desired_state: SwerveModuleState):
         )
 
     def zero(self):
```
Comment (auscompgeek) [inline components/chassis.py:113]: Should this still be called `zero`, seeing as it doesn't actually zero? Should it be more like "synchronise motor sensor to encoder"?

File: components/chassis.py [hunk 5]
```diff
@@ -70,13 +73,16 @@ def __init__(
         self.target_angle = 0
 
     def get_angle(self) -> float:
-        # return self.hall_effect.getPosition()
-        return self.get_motor_angle() * self.STEER_SENSOR_TO_RAD
+        """Gets steer angle from absolute encoder"""
+        # encoder returns in 0-1
+        return self.encoder.getPosition() * math.tau
```
Comment (auscompgeek) [inline components/chassis.py:78]: Aren't you already doing this in the encoder object?

File: components/chassis.py [hunk 6]
```diff
@@ -221,19 +226,12 @@ def drive_local(self, x, y, z):
     def _drive(self, chassis_speeds):
         self.desired_states = self.kinematics.toSwerveModuleStates(chassis_speeds)
         for state, module in zip(self.desired_states, self.modules):
-            # new_state = SwerveModuleState.optimize(state, module.get_rotation())
-            module.set(state)
+            new_state = SwerveModuleState.optimize(state, module.get_rotation())
+            module.set(new_state)
```
Comment (auscompgeek) [inline components/chassis.py:230]: nit: make it easier to comment out the optimize call
```suggestion
            state = SwerveModuleState.optimize(state, module.get_rotation())
            module.set(state)
```

File: components/chassis.py [hunk 7]
```diff
@@ -74,16 +76,15 @@ def __init__(
...
 
     def get_motor_angle(self) -> float:
         """Gets steer angle from motor's integrated relative encoder"""
-        return self.steer.getSelectedSensorPosition()
+        return self.steer.getSelectedSensorPosition() * self.STEER_SENSOR_TO_RAD
 
     def get_rotation(self) -> Rotation2d:
-        """Absolute steer position as rotation2d"""
-        return Rotation2d(self.get_angle())
+        """Relative steer position as rotation2d"""
```
Comment (auscompgeek) [inline components/chassis.py:86]:
```suggestion
        """Get the angle of the module."""
```

**Review by auscompgeek — APPROVED** [review]
lgtm

**Pull request merged**